lib/elixir/src/elixir_env.erl

```diff
@@ -138,10 +139,14 @@ is_compiler_var([$_]) -> true;
 is_compiler_var([Var | Rest]) when Var =:= $_; Var >= $A, Var =< $Z -> is_compiler_var(Rest);
 is_compiler_var(_) -> false.
 
-format_error({unused_var, Name}) ->
-  case atom_to_list(Name) of
-    "_" ++ _ ->
-      io_lib:format("unknown compiler variable \"~ts\" (expected one of __MODULE__, __ENV__, __DIR__, __CALLER__, __STACKTRACE__)", [Name]);
-    _ ->
-      io_lib:format("variable \"~ts\" is unused (if the variable is not meant to be used, prefix it with an underscore)", [Name])
-  end.
+format_error({unused_var, Name, VariableWithSameNameExistsInContext}) when is_atom(Name) ->
+  format_error({unused_var, atom_to_list(Name), VariableWithSameNameExistsInContext});
+
+format_error({unused_var, "_" ++ _ = Name, _VariableWithSameNameExistsInContext}) ->
+  io_lib:format("unknown compiler variable \"~ts\" (expected one of __MODULE__, __ENV__, __DIR__, __CALLER__, __STACKTRACE__)", [Name]);
+
+format_error({unused_var, Name, false}) ->
+  io_lib:format("variable \"~ts\" is unused (if the variable is not meant to be used, prefix it with an underscore)", [Name]);
+
+format_error({unused_var, Name, true}) ->
+  io_lib:format("variable \"~ts\" is unused (there is a variable with the same name in the context, did you mean to use the pin-operator?)", [Name]).
```

Comment:
Should the message include the `underscore` hint from the non-overridden warning as well?

Reply:
What do you think about: (there is a variable with the same name in the context, use the pin operator (^) to match on it or prefix this variable with underscore if it is not meant to be used) ?

Reply:
I like it :+1: